### PR TITLE
Use Course Offering and Course Version for Assignment Dropdown and Version Selector

### PR DESCRIPTION
**Pull request by dmcavoy**
This PR updates the format of information for: 
- Assignment Dropdown (Homepage and Teacher Dashboard)
Screen Shot 2022-03-30 at 3 43 10 PM

- Version Selector (Course and Unit Overview)
Screen Shot 2022-03-30 at 3 44 03 PM

The new format relies on Course Offering, Course Version and Units to summarize all the Course Offerings that are available for a user and use the information to populate the assignment dropdown when creating a section or the version selector on a course. The summarizing of this information was done ahead of time in other PRs so this PR just switches over to using that new set of information. Check out `assignable_course_offerings_info` in course_offering.rb to learn more about that summary.

In addition to populating the information based on Course Offering, Course Version and Units we are now also assigning courses based on course_verison_id and unit_id. So when `create` and `update` are called for a section we send the `course_version_id` and `unit_id` (if applicable). 

We also assign courses/units from the Course and Unit Overview page so some updates were made to get the correct assignment information on those pages as well. 
Screen Shot 2022-03-30 at 3 43 45 PM

More details about the updates being made here can be found in the Mini Spec linked below.

## Tips for Reviewing this PR

Consider viewing the changes in split mode. There is a lot of old logic that had to be removed and it can be hard to read the way github is formatting the changes. 

Consider starting with these files:
- AssignmentSelctor.jsx 
- teacherSectionsRedux.js

Those two files are the heart of a lot of the front end changes. Understanding those two files will allow you to understand some of the smaller changes in other places. 

Then take a look at sections_controller.rb. This is where we have updated from create and update getting `course_id` and `script_id` to `course_version_id` and `unit_id`. 

## Links

[Jira](https://codedotorg.atlassian.net/browse/PLAT-1584)
[Eng Plan - Self Paced Pl](https://docs.google.com/document/d/1V61xONU0-mleMEEdHWNxhvZtYIPRCB31mT_hbUXusCQ/edit#heading=h.d0b9vmlcrsox)
[Mini Spec - Course Assignment Dropdown Updates](https://docs.google.com/document/d/11-dtCFuSmST_vP2MdkH0H0lb2e5AGFu1MxiWSRsnkns/edit#)

## Testing story

- [Bug Bashed](https://docs.google.com/document/d/1sWRTYZ1U557DG9-IJz2FdCNdSf4WjpcWxNsiViRsT3Q/edit#) on Friday March 25th
- Existing UI Tests (course_versions.feature/assignment_on_overview_pages.feature/script_overview.feature/teacher_dashboard_assessment1.feature/teacher_dashboard_assessment2.feature/teacher_homepage.feature/etc.)
- Updated UI Tests
- Existing Unit Tests(sections_controller_test.rb/teacherSectionsReduxText.js/AssignmentSelectorTest.js/etc.)
- Updated Unit Tests

## Follow-up work

- Assignment flow for PL courses

### CODE REVIEW

File: apps/src/templates/teacherDashboard/teacherSectionsRedux.js [hunk 1]
```diff
@@ -1155,16 +1100,37 @@ export function isSaveInProgress(state) {
   return getRoot(state).saveInProgress;
 }
 
+function assignedUnit(state) {
+  const {sectionBeingEdited, courseOfferings} = getRoot(state);
+
+  return courseOfferings[sectionBeingEdited.courseOfferingId]?.course_versions[
```
Comment (dmcavoy) [inline apps/src/templates/teacherDashboard/teacherSectionsRedux.js:1106]: The assigned unit could also be the course version

File: apps/src/templates/teacherDashboard/shapes.jsx [hunk 2]
```diff
@@ -69,50 +72,34 @@ export const assignmentShape = PropTypes.shape({
   supported_locale_codes: PropTypes.arrayOf(PropTypes.string)
 });
 
-// An assignment family is a collection of versions of a course or script like
```
Comment (dmcavoy) [inline apps/src/templates/teacherDashboard/shapes.jsx:72]: Is assignmentShape still used?

File: dashboard/app/controllers/api/v1/sections_controller.rb [hunk 3]
```diff
@@ -92,7 +90,7 @@ def update
     # TODO: (madelynkasula) refactor to use strong params
     fields = {}
     fields[:course_id] = set_course_id(course_id)
-    fields[:script_id] = set_script_id(script_id)
+    fields[:script_id] = set_script_id(unit_id)
```
Comment (dmcavoy) [inline dashboard/app/controllers/api/v1/sections_controller.rb:93]: Come back and looks at set_script_id and set_course_id

**Comment by dmcavoy** [comment]
Need to order years in reverse order and need to show date range instead of just version year

Screen Shot 2022-03-15 at 2 49 08 PM

**Comment by dmcavoy** [comment]
Automatically assign the first unit in the course when pick course

Automatically read instructions aloud to students? (Only certain courses and web browsers) Not showing up for AIML standalone

Turn on Lesson Extras? (Only certain courses and units) not showing up for Course E

Decide later does not keep decide later in the dropdown

On Unit Overview page for Course A - Recommended does not show in the version selector

Assign to section button not working on unit overview page

**Comment by dmcavoy** [comment]
Unassign with assign button did not work

**Comment by dmcavoy** [comment]
Showing unit under unit

Screen Shot 2022-03-16 at 11 02 58 AM

**Comment by dmcavoy** [comment]
Make sure units are in correct order based on order in unit group

**Comment by dmcavoy** [comment]
Decide later does not keep decide later in the dropdown
Make sure units are in correct order based on order in unit group

**Comment by dmcavoy** [comment]
Course 1 and 20-hour are not assigning...

**Comment by dmcavoy** [comment]
Finish getting course and unit overview assigning working

**Comment by dmcavoy** [comment]
Assignment does not stick on course overview page

**Comment by dmcavoy** [comment]
Assignment does not stick on course overview page
Decide later does not keep decide later in the dropdown
Make sure units are in correct order based on order in unit group

Students can not see version selector on unit and course overview pages

File: dashboard/test/ui/features/learning_platform/teacher_homepage.feature [hunk 4]
```diff
@@ -14,16 +14,6 @@ Feature: Using the teacher homepage sections feature
     When I create a new section and go home
     Then the section table should have 2 rows
 
-  Scenario: Loading teacher homepage with course experiment enabled
```
Comment (dmcavoy) [inline dashboard/test/ui/features/learning_platform/teacher_homepage.feature:17]: Based on conversation with Dave and Tess, we are removing the ability to assign alternative units.

**Comment by dmcavoy** [comment]
Double check levelbuilder is working correctly and was just adhoc bugs

**Comment by dmcavoy** [comment]
Note to Dani: Make sure courses and scripts that are hidden all show the *

**Comment by dmcavoy** [comment]
Note to Dani: Make sure everything has a course offering

**Comment by dmcavoy** [comment]
Note to Dani: Make sure courses and scripts that are hidden all show the *
Note to Dani: Make sure everything has a course offering

File: apps/src/code-studio/components/progress/UnitOverviewHeader.jsx [hunk 5]
```diff
@@ -127,17 +121,6 @@ class UnitOverviewHeader extends Component {
       versionWarningDetails = i18n.wrongCourseVersionWarningDetails();
     }
 
-    // Only display viewable versions in script version dropdown.
-    const filteredVersions = versions.filter(version => version.canViewVersion);
```
Comment (bethanyaconnor) [inline apps/src/code-studio/components/progress/UnitOverviewHeader.jsx:131]: Why don't we need to filter these anymore?

File: apps/src/templates/teacherDashboard/teacherSectionsRedux.js [hunk 6]
```diff
@@ -17,7 +17,6 @@ const USER_EDITABLE_SECTION_PROPS = [
   'ttsAutoplayEnabled',
   'participantType',
   'courseId',
-  'scriptId',
```
Comment (bethanyaconnor) [inline apps/src/templates/teacherDashboard/teacherSectionsRedux.js:20]: Just checking: this is redundant with `unitId` right?

File: apps/test/unit/templates/courseOverview/CourseOverviewTest.js [hunk 7]
```diff
@@ -168,68 +169,25 @@ describe('CourseOverview', () => {
...
-          isStable: true,
-          locales: [],
-          localeCodes: []
-        }
-      ];
       const wrapper = shallow(
         <CourseOverview
           {...defaultProps}
-          versions={versions}
+          versions={courseOfferings['2'].course_versions}
```
Comment (bethanyaconnor) [inline apps/test/unit/templates/courseOverview/CourseOverviewTest.js:175]: These indices are a bit confusing to me. How did you decide which course offering to use?

File: dashboard/app/controllers/api/v1/sections_controller.rb [hunk 8]
```diff
@@ -260,23 +260,26 @@ def find_follower
   end
 
   def get_course_and_unit
-    return head :forbidden if current_user.nil?
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/api/v1/sections_controller.rb:263]: Why remove this check?

File: dashboard/test/controllers/api/v1/sections_controller_test.rb [hunk 9]
```diff
@@ -469,13 +469,14 @@ class Api::V1::SectionsControllerTest < ActionController::TestCase
   end
 
   [CSP_COURSE_NAME, CSP_COURSE_SOFT_LAUNCHED_NAME].each do |existing_unit_group_name|
-    test "can create with a course id but no script id - #{existing_unit_group_name}" do
+    test "can create with a course as course version but no unit selected - #{existing_unit_group_name}" do
       existing_unit_group = UnitGroup.find_by(name: existing_unit_group_name)
+      CourseOffering.add_course_offering(existing_unit_group)
```
Comment (bethanyaconnor) [inline dashboard/test/controllers/api/v1/sections_controller_test.rb]: Why is this needed?

File: dashboard/test/controllers/api/v1/sections_controller_test.rb [hunk 10]
```diff
@@ -485,41 +486,42 @@ class Api::V1::SectionsControllerTest < ActionController::TestCase
     end
   end
 
-  test 'cannot assign an invalid course id' do
+  # Check this
```
Comment (bethanyaconnor) [inline dashboard/test/controllers/api/v1/sections_controller_test.rb]: Forgot to delete?

File: dashboard/test/controllers/api/v1/sections_controller_test.rb [hunk 11]
```diff
@@ -485,41 +486,42 @@ class Api::V1::SectionsControllerTest < ActionController::TestCase
...
+  test 'cannot assign an invalid course version id' do
     sign_in @teacher
     post :create, params: {
       login_type: Section::LOGIN_TYPE_EMAIL,
-      course_id: @beta_unit_group.id, # Not CSP or CSD
+      course_version_id: -1,
```
Comment (bethanyaconnor) [inline dashboard/test/controllers/api/v1/sections_controller_test.rb]: This test's name doesn't seem to match this but `@beta_unit_group` looks like it's a valid unit group, just not an assignable one. Is there a reason this is changed?

**Review by bethanyaconnor — APPROVED** [review]
Left a few questions but overall looks good to me, especially with the successful bug bash. I'd love for at least one other person to look at it though, as I'm not super familiar with the assignment dropdown and this is a pretty big PR.

SUPER excited to see this happening!

File: apps/src/code-studio/components/progress/UnitOverviewHeader.jsx (hunk 5 above)
Comment (dmcavoy) [reply apps/src/code-studio/components/progress/UnitOverviewHeader.jsx:131]: We do it all on the backend before we pass the information down as part of `assignable_course_offerings_info`

File: apps/src/templates/teacherDashboard/teacherSectionsRedux.js (hunk 6 above)
Comment (dmcavoy) [reply apps/src/templates/teacherDashboard/teacherSectionsRedux.js:20]: Yup - Just moving over to unitId

File: apps/test/unit/templates/courseOverview/CourseOverviewTest.js (hunk 7 above)
Comment (dmcavoy) [reply apps/test/unit/templates/courseOverview/CourseOverviewTest.js:175]: Different courseOfferings have different properties I was trying to test out. Some have one course_version some have multiple. Some have content roots that are Script and some are UnitGroups. So for this one I wanted something with two course_versions so I used the CSD course offering test data

File: apps/src/templates/teacherDashboard/AssignmentSelector.jsx [hunk 12]
```diff
@@ -278,28 +197,28 @@ export default class AssignmentSelector extends Component {
...
 
-    const assignmentFamiliesByCategory = categorizeAssignmentFamilies(
-      assignmentFamilies
+    const orderedUnits = _.orderBy(selectedCourseVersion?.units, 'position');
+
+    const filteredCategories = _.filter(
+      Object.keys(CourseOfferingCategories),
+      function(category) {
+        return courseOfferingsByCategories[category];
+      }
```
Comment (davidsbailey) [inline apps/src/templates/teacherDashboard/AssignmentSelector.jsx:227]: makes sense per offline discussion, but would be helpful to comment what this block is doing and why.

File: apps/test/unit/templates/teacherDashboard/AssignmentSelectorTest.js [hunk 13]
```diff
@@ -164,260 +72,154 @@ describe('AssignmentSelector', () => {
...
-      courseId: null,
-      scriptId: null
-    });
-  });
-
-  it('includes uncoursed script if section is assigned that script currently', () => {
-    // Eventually we don't expect this to be a plausible case, but until we do
-    // our migration, it's possible for a section to be assigned to a script that
-    // is in a course, and we want to make sure this still shows up in our
-    // dropdown.
```
Comment (davidsbailey) [inline apps/test/unit/templates/teacherDashboard/AssignmentSelectorTest.js:315]: I think it would be worth either (1) confirming that we do not have production data in this state, or (2) testing that we still do something reasonable in this case (anything without a js error is ok with me).

**Review by davidsbailey — APPROVED** [review]
Awesome work Dani! Thanks for the walkthrough. Code changes look good assuming we don't have any sections-with-units-without-courses in the DB -- if not, let's discuss what the risks are there before proceeding.

File: apps/test/unit/templates/teacherDashboard/AssignmentSelectorTest.js (hunk 13 above)
Comment (davidsbailey) [reply apps/test/unit/templates/teacherDashboard/AssignmentSelectorTest.js:315]: per offline convo, this is implemented / tested on the backend now, so its ok to remove the test here.
